File: Python/Problem_2610.py
from typing import DefaultDict, List

#   Problem - 2610: https://leetcode.com/problems/convert-an-array-into-a-2d-array-with-conditions/
#   Python Solution!
class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]:
        numCount = DefaultDict(int)
        res = []

        for num in nums:
            row = numCount[num]

            if row == len(res):
                res.append([])

            res[row].append(num)
            numCount[num] += 1

        return res
    

if __name__=="__main__":
    sol = Solution()

    print(sol.findMatrix([1,3,4,1,2,3,1]))
    print(sol.findMatrix([1,2,3,4]))
    print(sol.findMatrix([2,1,1]))

File: Cpp/Problem_2610.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

// Problem - 2610: https://leetcode.com/problems/convert-an-array-into-a-2d-array-with-conditions/
// C++ Solution!
class Solution {
public:
	vector<vector<int>> findMatrix(vector<int>& nums) {
		unordered_map <int, int> numCount;

		vector<vector<int>> res;
		for (int num : nums) {
			int row = numCount[num];

			if (row == res.size()) {
				res.push_back({ });
			}
			res[row].push_back(num);
			numCount[num]++;
		}

		return res;
	}
};

void printVector(vector<vector<int>> res) {
	for (vector<int> row : res) {
		for (int val : row) {
			cout << val << " ";
		}
		cout << endl;
	}
	cout << endl;
}

int main() {
	Solution sol;

	vector <int> input = { 1, 3, 4, 1, 2, 3, 1 };
	printVector(sol.findMatrix(input));
	input = { 2, 1, 1 };
	printVector(sol.findMatrix(input));

	return 0;
}